bundledata.go

```diff
@@ -336,7 +230,96 @@ type ApplicationSpec struct {
 	RequiresTrust bool `bson:"trust,omitempty" json:"trust,omitempty" yaml:"trust,omitempty"`
 }
 
-// ExpoExposedEndpointSpec describes the expose parameters for an application
+// maskedBundleData and bundleData are here to perform a way to normalize the
+// bundle data when unmarshalling via a codec.
+// By abusing the types we can prevent a recursive function call so that the
+// unmarshalling doesn't call itself.
+//
+// In reality this is so wrong in so many ways:
+//  1. Why has the model type got anything to do with how it's transferred over
+//     the wire to other consumables? The bundle data should have a package of
+//     wire protocols (DTOs) that can packed and unpacked into a bundle/charm
+//     model. That model should be pure!
+//  2. This should be a two step process, unmarshal and then normalize.
+type maskedBundleData BundleData
+
+type bundleData struct {
+	maskedBundleData `bson:",inline" yaml:",inline" json:",inline"`
+}
+
+// UnmarshalJSON implements the json.Unmarshaler interface.
+func (bd *BundleData) UnmarshalJSON(b []byte) error {
+	var in bundleData
+	if err := json.Unmarshal(b, &in); err != nil {
```

Comment:
I think we should detect the presence of services in bundle data and return an error to indicate they are not supported.

Reply:
Maybe... I'm unsure about this.